tool/teleport/common/teleport.go

```diff
@@ -489,6 +515,271 @@ func OnStart(clf config.CommandLineFlags, config *service.Config) error {
+	}
+
+	addr, err := utils.ParseAddr(clf.ProxyServer)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	privateKey, sshPublicKey, tlsPublicKey, err := GenerateKeys()
+	if err != nil {
+		return trace.Wrap(err, "unable to generate new keypairs")
+	}
+
+	ctx := context.Background()
+	hostname, err := getAWSInstanceHostname(ctx)
+	if err != nil {
+		var hostErr error
+		hostname, hostErr = os.Hostname()
+		if hostErr != nil {
+			return trace.NewAggregate(err, hostErr)
+		}
+	}
+
+	// TODO(amk) get uuid from a cli argument once agentless inventory management is implemented to allow tsh ssh access via uuid
+	uuid, err := uuid.GenerateUUID()
+	_ = err
+
+	principals := []string{uuid}
+	for _, principal := range strings.Split(clf.AdditionalPrincipals, ",") {
+		if principal == "" {
+			continue
+		}
+		principals = append(principals, principal)
+	}
+
+	registerParams := auth.RegisterParams{
+		Token:                clf.AuthToken,
+		AdditionalPrincipals: principals,
+		JoinMethod:           types.JoinMethod(clf.JoinMethod),
+		ID: auth.IdentityID{
+			Role:     types.RoleNode,
+			NodeName: hostname,
+			HostUUID: uuid,
+		},
+		ProxyServer:        *addr,
+		PublicTLSKey:       tlsPublicKey,
+		PublicSSHKey:       sshPublicKey,
+		GetHostCredentials: client.HostCredentials,
+		FIPS:               clf.FIPS,
+	}
+
+	if clf.FIPS {
+		registerParams.CipherSuites = defaults.FIPSCipherSuites
+	}
+
+	certs, err := auth.Register(registerParams)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	identity, err := auth.ReadIdentityFromKeyPair(privateKey, certs)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	client, err := authenticatedUserClientFromIdentity(ctx, clf.FIPS, *addr, identity)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	cas, err := client.GetCertAuthorities(ctx, types.UserCA, false)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	var userCA []byte
+	for _, ca := range cas {
+		for _, key := range ca.GetActiveKeys().SSH {
+			userCA = append(userCA, key.PublicKey...)
+			userCA = append(userCA, byte('\n'))
+		}
+	}
+
+	fmt.Printf("Writing Teleport keys to %s\n", clf.OpenSSHKeysPath)
+	if err := writeKeys(clf.OpenSSHKeysPath, privateKey, certs, userCA); err != nil {
+		return trace.Wrap(err)
+	}
+
+	fmt.Println("Updating OpenSSH config")
+	if err := updateSSHDConfig(clf.OpenSSHKeysPath, clf.OpenSSHConfigPath); err != nil {
+		return trace.Wrap(err)
+	}
+
+	fmt.Println("Restarting the OpenSSH daemon")
+	if err := restartSSHD(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+const (
+	teleportKey    = "teleport"
+	teleportCert   = "teleport-cert.pub"
+	teleportUserCA = "teleport_user_ca.pub"
+)
+
+func writeKeys(sshdConfigDir string, private []byte, certs *proto.Certs, userCA []byte) error {
+	if err := os.WriteFile(filepath.Join(sshdConfigDir, teleportKey), private, 0600); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := os.WriteFile(filepath.Join(sshdConfigDir, teleportCert), certs.SSH, 0600); err != nil {
```

Comment:
If this write fails for any reason, we will leave the `teleportKey` on the disk. The same apply to `teleportUserCA` (then `teleportCert` and `teleportKey` will be left). Should we remove it on the write error?

Reply:
@r0mant Any thoughts? I think that if the installation fails, we should leave the machine in the same state as if the installation would never happen. Otherwise, we may encounter weird errors on the second attempt.

Reply:
I agree, we should clean stuff up.

Reply:
But the code should tolerate files potentially being already present there just in case we failed to clean them up or something.

Reply:
> tolerate files potentially being already

I think this is more complicated in practice. First of all, we already recommend installing our certificates in `/etc/ssh` for proxy recording https://goteleport.com/docs/server-access/guides/recording-proxy-mode/
If we decide to "ignore already existing files", we may override some existing files not created by Teleport. TBH now when I think about it, I don't see why we would keep our certificates or any other files outside of `/etc/teleport`. I think that the only file that we need to modify outside of that directory is sshd config.
This is what I would do:

1. Create `/etc/teleport` if doesn't exist
2. Create `/etc/teleport/agentless` directory if doesn't exist. If the directory already exists, remove it (that should cover "failed previous installations" case).
3. Write all certs and keys inside `/etc/teleport/agentless`.
4. Create a backup of the sshd config (`/etc/ssh/sshd_config.teleport` maybe?).
5. Update the original sshd config and test it.
6. Restart sshd daemon.

In case of any failures, we would need to just remove `/etc/teleport/agentless`. This directory should also be added to our RPM/DEB packages to be removed on "uninstall" comments. In case of sshd failures, we should still be able to revert all changes as we will have the backed-up config.

That being said, this process (current and described here) will fail if a machine has SELinux enabled as Teleport process even running as a root won't have access to modify sshd config. I don't think that SELinux support should be included in the first implementation anyways. I just want to mention that we should be prepared to see errors related to it.